File: src/test/java/com/ealanta/productapp/label/LabelledProductSerializationtTest.java
package com.ealanta.productapp.label;

import com.ealanta.productapp.label.LabelledProduct;
import com.ealanta.productapp.misc.TestUtils;
import com.ealanta.productapp.price.PriceInfo;
import com.ealanta.productapp.product.ColorSwatch;
import com.ealanta.productapp.product.Product;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import org.json.JSONException;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.skyscreamer.jsonassert.JSONAssert;
import org.skyscreamer.jsonassert.JSONCompareMode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.Resource;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.util.StreamUtils;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;

@RunWith(SpringRunner.class)
public class LabelledProductSerializationtTest {

    private Logger LOGGER = LoggerFactory.getLogger(LabelledProductSerializationtTest.class);

    @Value("classpath:/output/labelledProduct1.json")
    private Resource labelledProduct;

    private String labelledProductJson;

    @Before
    public void setup() throws IOException, JSONException {
        this.labelledProductJson = StreamUtils.copyToString(labelledProduct.getInputStream(), StandardCharsets.UTF_8);
        LOGGER.debug(this.labelledProductJson);

    }
    @Test
    public void testDeserialize() throws IOException,JSONException {
        LabelledProduct product = new LabelledProduct();
        product.setProductId("productID");
        product.setTitle("TITLE");
        product.setNowPrice("£12.34");
        PriceInfo info = new PriceInfo();
        info.setNowPrice(TestUtils.getMoney("21.12"));
        product.setPriceInfo(info);
        product.setPriceLabel("THIS IS THE PRICE LABEL");

        ColorSwatch sw1 = new ColorSwatch();
        sw1.setRgbColor("RGB1");
        sw1.setColor("COLOR1");
        sw1.setBasicColor("BASIC1");
        sw1.setSkuId("SKU1");

        ColorSwatch sw2 = new ColorSwatch();
        sw2.setRgbColor(null);
        sw2.setColor("COLOR2");
        sw2.setBasicColor("BASIC2");
        sw2.setSkuId("SKU2");

        product.getColorSwatches().addAll(Arrays.asList(sw1,sw2));

        ObjectMapper mapper = new ObjectMapper();
        mapper.enable(SerializationFeature.INDENT_OUTPUT);

        String productJson = mapper.writeValueAsString(product);

        JSONAssert.assertEquals(labelledProductJson, productJson, JSONCompareMode.LENIENT);

    }
}

File: src/src/test/kotlin/com/ealanta/productapp/label/LabelledProductSerializationtTest.kt
package com.ealanta.productapp.label;

import com.ealanta.productapp.domain.LabelledProduct
import com.ealanta.productapp.domain.PriceInfo;
import com.ealanta.productapp.domain.ColorSwatch;
import com.ealanta.productapp.utils.TestUtils
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import mu.KotlinLogging
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.skyscreamer.jsonassert.JSONAssert;
import org.skyscreamer.jsonassert.JSONCompareMode;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.Resource;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.util.StreamUtils;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import kotlin.properties.Delegates

@RunWith(SpringRunner::class)
class LabelledProductSerializationtTest {

    private val logger = KotlinLogging.logger {}

    @Value("classpath:/output/labelledProduct1.json")
    lateinit var labelledProduct:Resource

    var labelledProductJson:String by Delegates.notNull()

    @Before
    fun setup() {
        this.labelledProductJson = StreamUtils.copyToString(labelledProduct.getInputStream(), StandardCharsets.UTF_8);
        logger.debug(this.labelledProductJson);

    }
    @Test
    fun testDeserialize() {
        val product = LabelledProduct()
        product.productId = "productID";
        product.title = "TITLE";
        product.nowPrice = "£12.34";
        val info = PriceInfo();
        info.nowPrice = TestUtils.getMoney("21.12");
        product.priceInfo = info;
        product.priceLabel = "THIS IS THE PRICE LABEL";

        val sw1 = ColorSwatch();
        sw1.rgbColor = "RGB1";
        sw1.color = "COLOR1";
        sw1.basicColor = "BASIC1";
        sw1.skuId = "SKU1";

        val sw2 = ColorSwatch()
        sw2.rgbColor = null;
        sw2.color = "COLOR2";
        sw2.basicColor = "BASIC2";
        sw2.skuId = "SKU2";

        product.colorSwatches.addAll(Arrays.asList(sw1,sw2));

        val mapper = jacksonObjectMapper().enable(SerializationFeature.INDENT_OUTPUT)

        val productJson:String? = mapper.writeValueAsString(product);
        println(productJson)

        JSONAssert.assertEquals(labelledProductJson, productJson, JSONCompareMode.LENIENT);

    }
}
